Micro Lava Starling

High

# Malicious owner will not be able to cancel the stream of their noun(`StreamEscrow::cancelStream`)

### Summary

Call to arbitrary address in `StreamEscrow::cancelStream` may fail as the address may be malicious and can have missing receive and fallback function by mistake or on purpose
Permalink : https://github.com/sherlock-audit/2024-11-nounsdao/blob/main/nouns-monorepo/packages/nouns-contracts/contracts/StreamEscrow.sol#L182

### Root Cause

The choice to transfer remaining money back to owner of noun on cancellation of stream is a mistake since the call can fail(by mistake or on purpose) , and the cancellation wouldn't go through

### Internal pre-conditions

1. A user who is the owner of a noun needs to start a stream
2. The owner needs to have a missing or malicious fallback and receive
3. Then the owner needs to call `cancelStream` , which would fail(revert)

### External pre-conditions

_No response_

### Attack Path

1. A user who is the owner of a noun needs to start a stream
2. The owner needs to have a missing or malicious fallback and receive
3. Then the owner needs to call `cancelStream` , which would fail(revert)

### Impact

The owner of an escrow stream may not be able to cancel their stream , and will be stuck with the noun even if they dont want it.

### PoC

In this PoC , I have simply made the  `fallback` and `receive` function to revert , but keep in mind the owner may have these functions wrongly implemented and/or missing also. The point is it is extra work and caution on the user's side to correctly implement their `fallback` and `receive` functions , which won't always be the case and this is where the `cancelStream` function will fail

In `StreamEscrow.t.sol`::

1. Add the following contract 

```solidity
    contract MaliciousUser {
    fallback() external payable{
        revert();
    }
    receive() external payable{
        revert();
    }
}
```

2. Add the following test to `CancelStreamTest` contract in `StreamEscrow.t.sol`

```solidity
    function test_CannotCancelWithMaliciousUser() public {
        // forward 5 days, quarter way through the stream
        for (uint i; i < 5; i++) {
            forwardOneDay();
        }

        MaliciousUser malUser = new MaliciousUser();
        vm.prank(user);
        nounsToken.transferFrom(user, address(malUser), 1);

        assertEq(ethRecipient.balance, 2.5 ether);
        assertEq(nounsToken.ownerOf(1), address(malUser));

        // cancel stream
        vm.prank(address(malUser));
        nounsToken.approve(address(escrow), 1);
        vm.prank(address(malUser));
        vm.expectRevert();
        escrow.cancelStream(1);
    }
```


### Mitigation

Use pull over push mechanism , i.e. , dont refund the owner right away in the same function , instead store the balances of people who call `cancelStream` using a mapping , and create a withdraw function , where people can withdraw their funds if they are eligible for it.